src/ir_passes/new_mono.rs

```diff
@@ -0,0 +1,868 @@
+        let (comp, params) = self.pass.should_process(*comp, conc_params);
+        let new_inst = ir::Instance {
+            comp,
+            params: params.into_iter().map(|n| self.base.num(n)).collect(),
+            info: self.info(info),
+        };
+
+        let new_idx = self.base.add(new_inst);
+
+        let inst_occurrences = self.inst_counter.get(&inst).unwrap();
+        self.inst_map.insert((inst, *inst_occurrences), new_idx);
+        new_idx
+    }
+
+    /// Monomorphize the port (owned by self.underlying) and add it to `self.base`, and return the corresponding index
+    fn port(&mut self, port: ir::PortIdx) -> ir::PortIdx {
+        if let Some(idx) = self.port_map.get(&port) {
+            println!("found {} -> {} in port_map", port, idx);
+            return *idx;
+        };
+        let ir::Port {
+            owner,
+            width,
+            live,
+            info,
+        } = self.underlying.get(port);
+
+        let ir::Liveness {
+            idx: underlying_idx,
+            ..
+        } = live;
+
+        let info = self.info(info);
+
+        // Add the new port so we can use its index in defining the correct Liveness
+        let new_port = self.base.add(ir::Port {
+            owner: owner.clone(),
+            width: *width,      // placeholder
+            live: live.clone(), // placeholder
+            info,
+        });
+
+        self.port_map.insert(port, new_port);
+        println!("inserted {} -> {} into port_map", port, new_port);
+
+        // Find the new port owner
+        let mono_owner = self.find_new_portowner(owner);
+
+        let ir::Liveness { idx, len, range } = live;
+
+        self.curr_port = Some(new_port);
+        let mono_liveness_idx = self.bundle_param(*idx);
+
+        let mut mono_liveness = ir::Liveness {
+            idx: mono_liveness_idx,
+            len: *len,            // placeholder
+            range: range.clone(), // placeholder
+        };
+
+        // Create a binding from old param -> new param
+        self.par_binding.insert(*underlying_idx, mono_liveness_idx);
+
+        // After making the new binding, re-monomorphize other parts of len in case they contained
+        // the param we replaced
+        let mono_width = self.expr(*width);
+        mono_liveness.len = self.expr(mono_liveness.len);
+        mono_liveness.range = self.range(&mono_liveness.range);
+
+        let port = self.base.get_mut(new_port);
+        port.live = mono_liveness; // update
+        port.width = mono_width; // update
+        port.owner = mono_owner; // update
+
+        new_port
+    }
+
+    /// Add `self.underlying`'s info to `self.base`. Nothing else needs to be done because all the constructs
+    /// we generate and add to `self.base` map to the same source-level info that they did in `self.underlying`
+    fn info(&mut self, info: &ir::InfoIdx) -> ir::InfoIdx {
+        let info = self.underlying.get(*info);
+        self.base.add(info.clone())
```

Comment:
`Info` is stored in an `IndexStore` I think which doesn't deduplicate things which means this will create a lot of copies of the same data over and over.